### ISSUE TITLE
Whole computer slows down while the program is open

### ISSUE THREAD

**Issue by radgeRayden** [dfranx/SHADERed#4]
I'm trying out SHADERed on an old(er) laptop with intel graphics, unfortunately on the simplest examples I can't do much on the computer while alt tabbed because it uses as much CPU time as possible (apparently). Setting the framecap to 60 didn't solve it.

Edit: after posting the issue I tried setting the vsync instead, it seems better but still taking a bit more cpu time than I'd like (for being minimized). I can't seem to be able todelete the issue so I'll leave it since it could be a legitimate issue still.

**Comment by dfranx** [comment]
Yeah, its because SHADERed wants to run as fast as possible and that is why it takes 100% of your CPU.

The frame cap built into the app only limits the FPS of the Preview window. I designed it this way so that (for example) when you want to see how your shader works when running at low FPS the rest of the app works at full speed and you can use it normally.

I guess some of the fixes could be:
- a checkbox that applies the frame cap to whole program (instead of Preview window only) when checked
- do some kind of sleep() to slow it down if SHADERed detects high cpu usage
- detect if the program is minimized/not focused and slow it down significantly using sleep functions so that its CPU usage is very low

Any suggestions?

**Comment by radgeRayden** [comment]
I'd be satisfied if it just slowed down when minimized, I suppose as long as it doesn't distort audio there's no reason for it to keep consuming resources.

**Comment by dfranx** [comment]
Mind waiting few weeks? MoonLight (the D3D library I wrote) can't detect if a window is minimized / not focused and I would rather spend time on porting to OpenGL & SDL than develop features for deprecated library.

**Comment by radgeRayden** [comment]
sure, I'm waiting for the openGL port anyways :) thanks!

**Comment by dfranx** [comment]
Hey @radgeRayden ! I just pushed a commit with some of the fixes mentioned above:
- SHADERed now runs at 30FPS when minimized (should significantly lower the CPU usage)
- there is an option to make SHADERed run at 60FPS when not focused (alt tabbed for example)
- there is an option to apply frame cap to whole program

The commit also includes the OpenGL port ;)
If you don't want to compile the program yourself you can wait for the binaries which will be released tomorrow/on Friday.

But please remember to test the fixes and see if any of them lower CPU usage! :D

**Comment by dfranx** [comment]
I have released the binaries - could you test the fixes?

**Comment by radgeRayden** [comment]
Hi, I just tested it. Unfortunately it didn't work too well.
GLSL Simple example doesn't show anything;
wrt the problem at hand, it seems it doesn't consume all the cpu anymore but when minimizing the window, the graphics black out and it has the computer chugging because now 100% of the GPU is being used (according to the task manager metric). This was with the ShaderToy example, using the intel graphics gpu.

**Comment by dfranx** [comment]
Oh wow, that is unfortunate.. Do you know if your GPU supports OpenGL 3.30 (core)? 

The program uses 100% of the GPU when minimized?? I want to fix all of your bugs/problems so badly... Please reply with as much as information as you can (exact GPU name, CPU, etc...)... There should also be log.txt in the same directory as the SHADERed executable... try checking if any error has occurred in there (try specifically checking the GLSL simple example which doesn't show anything in the preview window)...

**Comment by radgeRayden** [comment]
Ok, I tested a bit more. I have a switchable graphics setup, and this time I used my AMD card instead by switching it in the windows settings.
First, my hardware:
![image](https://user-images.githubusercontent.com/8618621/63561033-08de7c00-c51e-11e9-8e0e-e7fd9c001b1e.png)
![image](https://user-images.githubusercontent.com/8618621/63561115-55c25280-c51e-11e9-8609-36514851291e.png)

Each example seems to behave a bit differently, but I didn't find much difference between AMD and intel. I verified that it uses 100% gpu when focused, sometimes less when out of focus with the AMD (more beefy), but with AMD it didn't slow down the computer and as such didn't bother me much (although I'd like to avoid overheating).
As for the examples:
GLSLsimple is completely blank;
HLSLsimple crashes the program (!)
SHADERtoy/raymarching presents correctly but blank when out of focus
SHADERtoy/thedrivehome works as expected
procgradvis works as expected.
This is about as much as I tested. This was all AMD but if it's particularly important I can do more thorough tests.

log: https://gist.github.com/radgeRayden/8659bd1b3917319c504798749bcc8d3d

**Comment by radgeRayden** [comment]
I start to think the performance issues are corrected and I just observed an example that my gpu particularly struggled to deal with, if you want we can close this one and move the remaining problems to a new issue.

**Comment by radgeRayden** [comment]
With AMD GLSLsimple uses 100% GPU (for displaying nothing) when focused, and 50% when unfocused. With vsync on it uses 25% no matter what. Capping FPS to 60 and checking the "apply to whole application" box caps it at 50% for no vsync.

**Comment by dfranx** [comment]
Hey, just updated the release and now HLSLsimple shouldn't crash the program... Also, I kinda explored more whats causing the issue with the GLSLsimple being blank... Seems like some OpenGL extension or something is causing the issue... The GLSLsimple is blank on my laptop when running with the integrated Intel HD GPU, but it isn't blank when running with the dedicated NVIDIA GTX 840M... HLSL shaders work with both GPUs (even though HLSL is being translated to GLSL!!)... I am currently uncertain what causes this issue and will investigate it more in the next few days and hopefully fix it.

As with high GPU usage, I honestly have no idea how to fix that right now... On every PC I tried running SHADERed it has very high GPU usage (from 80% to 100%). I will try to fix this too but I am not sure when - I currently don't know how to fix it.

Also thank you for the screenshots and logs! The logs helped me fix the music visualized example in the ShaderToy directory after I have noticed this line:
> [20:54:19] (ERROR) Failed to load an audio file ../../../../../beat.wav
I forgot to change the path after I was done testing on my pc...

I will leave this issue open until I have a fix for the high GPU usage ;)

**Comment by dfranx** [comment]
Hey @radgeRayden could you please test the v1.1.1 I have just released? Tell me if GLSL shaders are working properly now and if HLSLsimple crashes please so that I can close issue #7 ! I think I have fixed those bugs!

**Comment by radgeRayden** [comment]
Can confirm no crashes or serious problems with the examples themselves anymore. Only the raymarching example still blacks out when out of focus. (tested on the x64 windows version).

**Comment by dfranx** [comment]
Great! I know what causes the raymarching example to go blank when out of focus... mouse position on shadertoy is clamped to [0,1] meanwhile SHADERed allows values greater than 1 or less than 0 - not sure right now if I should change it to ShaderToy style...

But anyway, I guess the only thing now left to do is lower down the GPU usage (if possible)

**Comment by dfranx** [comment]
Sorry, I don't think I can fix this on my side without overcomplicating the code. Commands being sent at high speeds to GPU is what causes the usage to be very high, at least I think so. I guess the fix would be to add some kind of render queue which, as I said, would complicate the code too much atm. The fix for this is just to use VSync option. It lowers the GPU usage significantly. Closing since the issue was firstly related to CPU usage when SHADERed is not focused (which I have fixed)!
